### DISCUSSION TITLE
Instalação do módulo fiscal l10n-brazil

### DISCUSSION THREAD

**Discussion by diegosayron** [OCA/l10n-brazil#2571] [Q&A] [answered]
Senhores, saudações.

Rodo o Odoo 16 no Ubuntu 24, instalei o módulo l10n-brazil no ambiente virtual:
`pip3 install -r l10n-brazil/requirements.txt`

No menu aplicativos, aparece como instalado, mas não me aparece nenhum menu fiscal/contábil, em lugar algum.
Alguém poderia informar se precisa de algo mais?

Em tempo, desenvolvo em C# para NF-e desde o início do projeto no Brasil, não tenho fluência em Python, mas se alguém puder orientar como poderia ajudar no projeto, backlog, etc, posso contribuir, se houver necessidade.
Grato!

**Comment by marcos-mendez** [comment]
https://t.me/odoobrazil

Entre aqui acho que é melhor

Em qua., 28 de jun. de 2023 às 20:18, DIEGO SAYRON ***@***.***>
escreveu:

**Comment by antoniospneto** [comment, marked as answer]
Olá, @diegosayron, seja bem-vindo à comunidade!

Os módulos de localização para a versão 16.0 do Odoo ainda não foram completamente migrados, apenas os mais básicos. Por isso, recomendo que você instale a versão 14.0, que é mais estável, madura e conta com todos os módulos. Usar a versão mais recente nem sempre é a melhor opção, especialmente para quem está iniciando. O desafio pode ser muito maior, pois você terá que lidar com muitos bugs ainda não resolvidos e funcionalidades/módulos que ainda não foram migrados.

Para quem está começando, recomendo a leitura do livro de Daniel Reis:  https://www.amazon.com/Odoo-Development-Essentials-development-applications-ebook/dp/B08KWLFW2H?ref_=ast_author_dp

Quanto ao Python, não acredito que será o seu maior desafio. Você vai se adaptando enquanto estuda o Odoo.

**Comment by laruzzo** [comment]
Pessoal, estou instalando na versão 14 e quando dou o comando
> pip install erpbrasil.edoc.pdf

Dá esse erro aqui, conseguem me dar uma ajuda? Obrigado

 Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      Traceback (most recent call last):
        File "<string>", line 36, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\f27602a\AppData\Local\Temp\pip-install-8avcx52a\sh_8a1674d067c9484da1632a746c25d6ac\setup.py", line 5, in <module>
          import sh
        File "C:\Users\f27602a\AppData\Local\Temp\pip-install-8avcx52a\sh_8a1674d067c9484da1632a746c25d6ac\sh.py", line 37, in <module>
          import fcntl
      ModuleNotFoundError: No module named 'fcntl'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

**Reply by rvalyi** [reply]
Olá @laruzzo, quais versões vc tem do Python e do pip?

Uma coisa que eu vejo é que vc está tentando no Windows. Não é uma boa, ninguém sério nem testa o Odoo no Windows (nem no Brasil nem na China)... E mesmo se vc superar esse problema com essa lib, pode ter certeza que logo logo vc vai encontrar problemas bem mais cabeludos que vão te dar mais trabalho de tentar resolver do que aprender o básico de um Ubuntu. O Python é portável na teoria, mas na prática a libc é e tudo que usa ela de forma indireta é diferente entre Linux e Windows então prós apps avançados ir de Windows é suicídio.

Enfim se puder responder sobre as versões do Python e do pip pode set que temos uma solução para essa lib.

**Reply by laruzzo** [reply]
Relamente, estou pesquisando aqui e me parece que o fcntl não existe para windows.
Estou usando:
pip 23.2.1 from C:\odoo14\python\lib\site-packages\pip (python 3.7)
Estou pensando em rodar via docker, será que dá certo? Meu servidor é ubuntu, mas estou testando em windows pra depois fazer o deploy.

**Reply by laruzzo** [reply]
vou abrir uma VM pra ubuntu aqui e tentar instalar
Obrigado pela dica

**Reply by antoniospneto** [reply]
> vou abrir uma VM pra ubuntu aqui e tentar instalar
> 
> Obrigado pela dica

Experimente o WSL do Windows, é uma boa opção para quem não quer abandonar o Windows de vez hehe

**Reply by laruzzo** [reply]
Consegui rodar no windows., Na verdade pip install erpbrasil.edoc.pdf dependia do sh e tem uma incompatibilidade com a versão mais nova. Declarei >pip install sh==1.13.0 e depois chamei >erpbrasil.edoc.pdf e funcionou

**Comment by marcos-mendez** [comment]
Tem gente que insiste no WSL é menos pior e funciona até que bem. Ubuntu ou
Debian são leves e funcionam ótimo

Em ter., 1 de ago. de 2023 às 14:53, Cezar Barreto ***@***.***>
escreveu:

**Comment by laruzzo** [comment]
Pessoal, consegui avançar aqui, mas para instalar o modulo l10n_br_fiscal_certificate está gerando um exception na linha 54 do arquivo 

File "c:\odoo14\server\addons\l10n-brazil-14.0\l10n-brazil-14.0\l10n_br_fiscal_certificate\hooks.py", line 54, in post_init_hook
    prepare_fake_certificate_vals()
Exception

Que é essa linha do código aqui, saberiam dizer o que devo fazer para resolver? O erro é por falta da tabela no postgree?

 try:
            for company in companies:
                l10n_br_fiscal_certificate_id = env["l10n_br_fiscal.certificate"]
                company.certificate_nfe_id = l10n_br_fiscal_certificate_id.create(
                    prepare_fake_certificate_vals()

**Reply by laruzzo** [reply]
Procurei essa tabela no postgree e não achei, talvez seja isso. Tem algum outro módulo ou script que tenho que instalar antes para gerar essa tabela? Obrigado

**Comment by wedsonlopes** [comment]
Boa tarde a todos, grande prazer! alguém já tem usado o l10n_brazil na versão 18 do odoo?

**Reply by rvalyi** [reply]
Ola @wedsonlopes até onde a gente sabe não. tem trabalho em curso para migrar para a 17.0 (so ver os PRs) mas ainda ta bem no inicio... A migração para a 18.0 nem começou.

Sempre existe alguns aventureiros tentando migrar de qualquer jeito fora da comunidade, porem a gente que é autor da maioria dos módulos desaconselha fortemente seguir esse caminho. Pois essas pessoas nunca fizeram ou conseguiram fazer um miserável PR na OCA se quer (as 10 pessoas do Brasil que ja fizeram a gente ja se conhece todo) e então basicamente se trata de pessoas que estão de bobeira como aparece todo ano antes de morrer alguns anos depois infelizmente.

Vc estaria fazendo ou seguindo de fato um fork que perderia qualquer chance de ter alguma sinergia com o OCA/l10n-brazil que nada menos é do que o maior dos 247 repos da OCA, sendo o módulo l10_br_fiscal o maior módulo dos 3000 módulos da OCA... Ou seja é uma baita furada andar de fora da comunidade...

Para vc ter uma noção, o repo OCA/l10n-brazil representa 10% do Odoo community inteiro tanto em termo de número de commits do que de linha de código. O Brasil sendo manicômio fiscal que é, não é para amador não...

Vale muito mais a pena vc deixar as últimas features da versão do ano com papai Noel e usar uma versão de 2 anos antes como a v16 (que ja foi a versão da vez 2 anos atrás alias) e migrar junto com a OCA.

Não deveria demorar muitos meses para a gente migrar os 15 módulos principais do projeto na 18.0. Então para empresa de serviço ou empresa do Simples que não é industria daqui alguns (5 ou 6 meses deveria rolar na 18.0). Mas se tiver mais pressa ou for empresa maior é melhor ir de 16.0 e migrar depois. Papo do fundador da Akretion, a empresa que lidera o projeto e co criou a OCA...

**Comment by rvalyi** [comment]
@wedsonlopes agora se vc quiser ajudar a accelerar o processo, tem una PRs relativamente simples para revisar para ajudar a dar um tapa em módulos essenciais antes de jogar para a 17.0 ou 18.0 como esse por exemplo: https://github.com/OCA/l10n-brazil/pull/3775

revisar isso ajudaria bastante a destravar...

**Comment by wedsonlopes** [comment]
show, vou tentar apoiar no que conseguir, abraço!
